### DISCUSSION TITLE
Size limits

### DISCUSSION THREAD

**Discussion by thesynman** [pybricks/support#278] [Q&A]
Hi,
Is there a guide to the maximum program size that will fit in the hub? I'm thinking of embarking on a complex robotics project using pybricks but I don't want to discover after weeks or months that my ideas won't fit. Thanks.
John

**Comment by laurensvalk** [comment]
Because MicroPython scripts get compiled to `mpy` byte code, it is tricky to give the size in terms of Python script size. However, I think we could show the compiled size in the Pybricks Code editor somehow. Perhaps during program download (we still have to add a progress indicator), or as information in a status pane. That way you could see the size of the scripts as it develops to get a sense of the required size.

So thanks for raising this issue.

In the mean time, perhaps we can help you if you want to share a bit more about your specific project.

Which hubs are you currently using? And what is the complexity you are looking at?

**Reply by thesynman** [reply]
Great idea to show the compiled size. I assume the capacity of each hub in terms of byte codes is known?

The project I have in mind isn't *that* complex, but it won't be trivial. I'm keen to develop an autonomous vehicle that can find and map objects (probably coloured plastic balls, at least to start), and then rearrange the objects in a prescribed pattern. But I'll start with something simpler :).

I'm going to start with the 4-port Technic hub, but I've not used any of the hubs before.

I recently retired after working as a professional software developer for more than 40 years, mostly building software tools including toolchains for embedded systems. I'm looking forward to contributing to the pybricks project.

**Reply by laurensvalk** [reply]
Thanks for getting back. That sounds like a cool project! 

The capacity is known for each hub indeed. At the moment, we allow the program to be at most half the heap size. The links below should provide some insight, as well as the heap size for each hub.

https://github.com/pybricks/pybricks-micropython/blob/411503b18aa4aa145f0c79014eb48f2ca321009b/bricks/technichub/mpconfigport.h#L11

https://github.com/pybricks/pybricks-micropython/blob/411503b18aa4aa145f0c79014eb48f2ca321009b/bricks/stm32/main.c#L66

**Reply by thesynman** [reply]
Very helpful. Thanks.

So at present 24KB is available to hold the byte codes. Is there a behind-the-scenes way right now to determine the compiled size of one of the example programs?

I was hoping there might be a way of storing the byte codes in flash rather than RAM.

**Reply by laurensvalk** [reply]
Programs are compiled with mpy-cross. We built a version into our web ide, but you can run it locally too. You can get it as a Python package via pip.

It also gets built from source when you build the firmware for the hubs. It sounds like you might be doing that anyway.

And yes, you can also save the script to flash, using the toggle in the settings menu to include your script during a firmware update. It still gets copied to RAM when it runs though, which is currently a MicroPython limitation.
